### PR TITLE
Saw

### PR DESCRIPTION
**Pull request by geralddejong**
Go through the changes, they're pretty interesting.

### CODE REVIEW

File: davinci-dome/src/DaVinci.ts [hunk 1]
```diff
@@ -20,11 +20,7 @@ export interface Joint {
 
 export interface Bar {
     index: number
-    jointA: Joint
-    jointB: Joint
-    jointC: Joint
-    jointD: Joint
-    planeJoint?: Joint
+    joints: Joint[]
```
Comment (geralddejong) [inline davinci-dome/src/DaVinci.ts:23]: jointA to jointD changed to joint[0] to joint[3], which allows for easily removing sawed off joints (below the plane)

File: davinci-dome/src/DaVinci.ts [hunk 2]
```diff
@@ -178,20 +177,26 @@ export function degreesToRadians(degree: number): number {
...
+            daVinciIntervals.push({
+                nodeIndexes: [bolt.jointA.index, bolt.jointB.index],
+                type: "type 2",
+            })
         })
-    })
     return {joints, daVinciIntervals}
 }
+
+function sawBar(bar: Bar, planeJoint: Joint): Bar {
```
Comment (geralddejong) [inline davinci-dome/src/DaVinci.ts:198]: Saw the bar means add the plane joint to either the front end or the back end (depending!), and then remove all joints with JointPosition.Below.

File: davinci-dome/src/Download.ts [hunk 3]
```diff
@@ -21,13 +21,20 @@ export interface DaVinciOutput {
 function extractNodeFile(output: DaVinciOutput): string {
     const rows: string[][] = []
     rows.push(["index", "x", "y", "z"])
-    output.joints.forEach((joint, index) => rows.push([
-        (index + 1).toFixed(0),
-        csvNumber(joint.point.x), csvNumber(joint.point.z), csvNumber(joint.point.y),
-    ]))
+    output.joints
+        .filter(joint => joint.position !== JointPosition.Below)
```
Comment (geralddejong) [inline davinci-dome/src/Download.ts:25]: ignore the joints which are below

File: davinci-dome/src/Parts.tsx [hunk 4]
```diff
@@ -26,9 +26,11 @@ const UP = new Vector3(0, 1, 0)
 const DOWN = new Vector3(0, -1, 0)
 
 export function BarBox({bar, renderSpec}: { bar: Bar, renderSpec: DaVinciSpec }): JSX.Element {
-    const length = bar.jointA.point.distanceTo(bar.jointD.point)
-    const unit = new Vector3().subVectors(bar.jointD.point, bar.jointA.point).normalize()
-    const midpoint = new Vector3().lerpVectors(bar.jointA.point, bar.jointD.point, 0.5)
+    const jointA = bar.joints[0]
+    const jointB = bar.joints[bar.joints.length -1]
```
Comment (geralddejong) [inline davinci-dome/src/Parts.tsx:30]: here we no longer know that there will be 4 joints.  there may have been joints sawed off, so we pick the last one.

File: davinci-dome/src/App.tsx [hunk 5]
```diff
@@ -49,9 +49,11 @@ function App(): JSX.Element {
         const radians = Math.abs(degreesToRadians(degrees))
         const result = daVinci(new Scaffold(frequency, radius, chiralityFromSpec(renderSpec)), radians, true)
         setDaVinciResult(daVinciToDome(result, renderSpec.planeHeight))
-        console.log("RENDER_SPEC", renderSpec.radius)
         setVersion(v => v + 1)
     }, [renderSpec])
+    const circleRadius = () => 1.5 * Math.sqrt(daVinciResult.joints.filter(j => j.position !== JointPosition.Below).reduce((r, {point}) => (
```
Comment (geralddejong) [inline davinci-dome/src/App.tsx:54]: dynamically calculate the circle radius so adjusts when the plane is raised

File: davinci-dome/src/App.tsx [hunk 6]
```diff
@@ -66,30 +68,31 @@ function App(): JSX.Element {
                     }
                     const position = face.normal.multiplyScalar(event.camera.position.length())
                     event.camera.position.copy(position)
-                }}>
+                }} position={new Vector3(0, renderSpec.planeHeight, 0)}>
```
Comment (geralddejong) [inline davinci-dome/src/App.tsx:71]: move the box to the plane

File: davinci-dome/src/App.tsx [hunk 7]
```diff
@@ -66,30 +68,31 @@ function App(): JSX.Element {
...
                 {daVinciResult.bolts
                     .filter(bolt => bolt.jointA.position === JointPosition.Above || bolt.jointB.position === JointPosition.Above)
                     .map((bolt, index) => (
-                    <BoltCylinder key={`bolt-${version}-#${index}`} bolt={bolt} renderSpec={renderSpec}/>
-                ))}
+                        <BoltCylinder key={`bolt-${version}-#${index}`} bolt={bolt} renderSpec={renderSpec}/>
+                    ))}
                 {daVinciResult.joints.filter(joint => joint.position === JointPosition.OnSurface).map((joint, index) => (
-                    <Ball key={`ball-${version}-#${index}`} position={joint.point} radius={renderSpec.radius / 50}/>
+                    <Ball key={`ball-${version}-#${index}`} position={joint.point} radius={renderSpec.barWidth}/>
```
Comment (geralddejong) [inline davinci-dome/src/App.tsx:90]: radius related to barWidth makes more sense

File: davinci-dome/src/App.tsx [hunk 8]
```diff
@@ -66,30 +68,31 @@ function App(): JSX.Element {
...
                 ))}
                 <pointLight position={[0, 10 * renderSpec.radius, 0]} color="white"/>
-                <PerspectiveCamera makeDefault={true} position={[renderSpec.radius * 3, 0, 0]}/>
-                <OrbitControls/>
+                <PerspectiveCamera makeDefault={true}
+                                   position={new Vector3(2 * circleRadius(), (renderSpec.planeHeight + renderSpec.radius) / 2, 0)}/>
+                <OrbitControls target={new Vector3(0, (renderSpec.planeHeight + renderSpec.radius) / 2, 0)}/>
```
Comment (geralddejong) [inline davinci-dome/src/App.tsx:95]: orbit controls targeted to the new box position, on the level of the plane

File: davinci-dome/src/App.tsx [hunk 9]
```diff
@@ -66,30 +68,31 @@ function App(): JSX.Element {
...
+                    ))}
                 {daVinciResult.joints.filter(joint => joint.position === JointPosition.OnSurface).map((joint, index) => (
-                    <Ball key={`ball-${version}-#${index}`} position={joint.point} radius={renderSpec.radius / 50}/>
+                    <Ball key={`ball-${version}-#${index}`} position={joint.point} radius={renderSpec.barWidth}/>
                 ))}
                 <pointLight position={[0, 10 * renderSpec.radius, 0]} color="white"/>
-                <PerspectiveCamera makeDefault={true} position={[renderSpec.radius * 3, 0, 0]}/>
-                <OrbitControls/>
+                <PerspectiveCamera makeDefault={true}
+                                   position={new Vector3(2 * circleRadius(), (renderSpec.planeHeight + renderSpec.radius) / 2, 0)}/>
```
Comment (geralddejong) [inline davinci-dome/src/App.tsx:94]: camera position dependent on circle radius

File: davinci-dome/src/App.tsx [hunk 10]
```diff
@@ -64,32 +66,35 @@ function App(): JSX.Element {
                     if (!face) {
                         return
                     }
-                    const position = face.normal.multiplyScalar(event.camera.position.length())
```
Comment (geralddejong) [inline davinci-dome/src/App.tsx:67]: Reorient relative to the new box position, on the plane

File: davinci-dome/src/App.tsx (hunk 5 above)
Comment (BuroCiviel) [reply davinci-dome/src/App.tsx:54]: nice!!

File: davinci-dome/src/DaVinci.ts (hunk 1 above)
Comment (BuroCiviel) [reply davinci-dome/src/DaVinci.ts:23]: oke

File: davinci-dome/src/Download.ts (hunk 3 above)
Comment (BuroCiviel) [reply davinci-dome/src/Download.ts:25]: Ik sta te trappelen om RFEM zo in te laden :)

**Review by BuroCiviel — COMMENTED** [review]
Great work!! Will try RFEM tonight!

**Pull request merged**